src/sqlfluff/core/rules/base.py

```diff
@@ -349,18 +349,37 @@ def _raw_slices_from_templated_slices(
 class RuleContext:
     """Class for holding the context passed to rule eval functions."""
 
-    segment: BaseSegment
-    parent_stack: Tuple[BaseSegment, ...]
-    siblings_pre: Tuple[BaseSegment, ...]
-    siblings_post: Tuple[BaseSegment, ...]
-    raw_stack: Tuple[RawSegment, ...]
-    memory: Any
+    # These don't change within a file.
     dialect: Dialect
-    path: Optional[pathlib.Path]
     fix: bool
     templated_file: Optional[TemplatedFile]
+    path: Optional[pathlib.Path]
 
-    @cached_property
+    # These change within a file.
+    segment: BaseSegment
+    parent_stack: Tuple[BaseSegment, ...] = field(default=tuple())
+    raw_segment_pre: Optional[RawSegment] = field(default=None)
+    raw_stack: Tuple[RawSegment, ...] = field(default=tuple())
+    memory: Any = field(default_factory=dict)
+    segment_idx: int = field(default=0)
+
+    @property
+    def siblings_pre(self) -> Tuple[BaseSegment, ...]:  # pragma: no cover
```

Comment:
Changed from a normal field to a property computed on the fly. Very few rules. use this or `siblings_post`, so don't waste time building these tuples unless needed.